### ISSUE TITLE
Cross-scheduler inconsistent result structure

### ISSUE THREAD

**Issue by davidhao3300** [dask/dask#8474]
**What happened**:

Due to a mistake in our code, we were persisting a dask dataframe in one scheduler, but then ran the compute while specifying threads scheduler. What was weird was that the computation returned a pandas Series of Futures instead of the expected DataFrame (or exception).

**What you expected to happen**:

I'm not sure what the best behavior is. Could be erroring on conflicting scheduler, running on the original scheduler, or running on the specified scheduler. Either way, the result structure should be consistent.

**Minimal Complete Verifiable Example**:

```python
from distributed import Client, LocalCluster
import pandas as pd
import dask.dataframe as dd

cluster = LocalCluster()
client=Client(cluster)

df = pd.DataFrame({'a': [1,2,3]})

ddf = dd.from_pandas(df, npartitions=3)

ddf = client.persist(ddf)

# Works, and returns a DF as expected
print(ddf.compute())

# Returns a pandas Series with a row per-partition with a future as the result
print(ddf.compute(scheduler="threads"))
```

**Anything else we need to know?**:

**Environment**:

- Dask version:
- Python version:
- Operating System:
- Install method (conda, pip, source):

**Comment by scharlottej13** [comment]
Thanks @davidhao3300 for opening up this issue! Would you feel comfortable opening up a PR, with the suggestion from @jsignell?

**Comment by jsignell** [comment]
@jrbourbeau pointed me to https://github.com/dask/dask/issues/7082 as a broader idea that would help with this specific issue. It is probably overkill if this is the main improvement, but keeping track of some metadata about persisted layers could make cases like this more obvious.

**Comment by jsignell** [comment]
Feel free to just open a pull request @ppsbs as soon as you start work.

**Comment by ppsbs** [comment]
Thanks, @jsignell !

I've worked a few hours on this task in the last 2 days and (based on my little experience with dask) I cannot see an unexpected behavior. As `ddf.compute(scheduler="threads")` runs asynchronously and `compute` always update the scheduler inside of `get_scheduler`, for me it is expected that it would return a future. And, on the other hand, as `ddf.compute()` runs synchronously, for me it's expected that it would get the result from the futures created by `client.persist(ddf)`. 

Or, maybe I haven't understood which would be the "objects that the scheduler can't understand" that you mentioned, @jsignell '^ ^   

Thank you!

cc @davidhao3300

**Comment by jsignell** [comment]
Once you are understand what is going on it might make sense, but it is not the behavior that a user would naively expect. This is not likely to be what the user intends to get back:

```python
In [6]: ddf.compute(scheduler="threads")
Out[6]: 
0    <Future: finished, type: pandas.core.frame.Dat...
1    <Future: finished, type: pandas.core.frame.Dat...
dtype: object
```

To close this issue, we would like to be able to at least give the user some warning that the operation is likely not what they intended.

**Comment by ppsbs** [comment]
I get your point @jsignell ! So, we could maybe raise an exception in the case where `compute` (and consequently `get_async`) are called with a dask graph whose result has already started being computed. Do you think that is a good path? Another option that I can think of is forcing `start_state_from_dask` to always convert cached futures to "results" (hence giving to `ddf.compute(scheduler="threads")` the same behavior of `ddf.compute()`), but I'm not sure of the impact of that modification.

**Comment by jsignell** [comment]
I think raising an exception or even just a warning is the right path.

**Comment by pavithraes** [comment]
With #8700, we raise a warning when we set `dask.compute(scheduler="<single-machine-scheduler>"`. 

It'll be nice to also raise a warning if the scheduler is set using `dask.config.set`.

Minimal example:

```python
import dask.dataframe as dd
from dask.distributed import Client

client = Client()

ddf = dd.DataFrame.from_dict({"x": [1, 2]}, npartitions=1)

ddf.compute(scheduler="sync") # UserWarning: Running on a single-machine scheduler when a distributed client is active might lead to unexpected results.

with dask.config.set(scheduler="sync"):
    ddf.compute() # No warnings

client.close()
```

@jsignell I'll re-open this, please let me know if it'll be better to open a fresh issue instead. :)

**Comment by fjetter** [comment]
I agree with the initial report and agree that there should be an error raised if something like this happens. However, I believe the warning we have right now is confusing. I consider it a perfectly valid use case to use a threaded scheduler and a distributed scheduler at the same time.
As a matter of fact, in the distributed repo we ended up having to ignore this warning frequently because this is indeed valid behavior

What should not be allowed is to use a different scheduler/executor on a collection that contains futures of another executor which is what happens in case of calling persist.
For instance, when using two different distributed clients, it is now allowed to mix the futures. This currently raises here https://github.com/dask/dask/blob/534d1d8bf7d7f4c5dc90085a47f4d4362e1312a7/dask/highlevelgraph.py#L389-L393
and I believe a similar logic should be in place for this issue. The warning that's currently in there is not sufficient.
